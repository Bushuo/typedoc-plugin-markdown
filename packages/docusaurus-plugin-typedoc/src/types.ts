export interface PluginOptions {
  id: string;
  docsRoot: string;
  out: string;
  sidebar: SidebarOptions | null;
  readmeTitle: string | undefined;
  globalsTitle?: string | undefined;
  plugin?: string[];
  readme?: string;
  disableOutputCheck?: boolean;
  entryPoints?: any;
  entryDocument: string;
  hideInPageTOC: boolean;
  hideBreadcrumbs: boolean;
  siteDir?: string;
}

export interface FrontMatter {
  id: string;
  title: string;
  slug?: string;
  sidebar_label?: string;
  hide_title?: boolean;
}

export interface SidebarOptions {
  fullNames: boolean;
  sidebarFile: string;
  indexLabel?: string;
  readmeLabel: string;
}

export interface Sidebar {
  [sidebarId: string]: SidebarItem[];
}

export interface SidebarCategory {
  type: string;
  label: string;
  items: SidebarItem[];
}

export type SidebarItem = SidebarCategory | string;
