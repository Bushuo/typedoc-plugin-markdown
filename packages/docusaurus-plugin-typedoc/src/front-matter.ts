import * as path from 'path';

import { BindOption } from 'typedoc';
import {
  getPageTitle,
  prependYAML,
} from 'typedoc-plugin-markdown/dist/utils/front-matter';
import { Component } from 'typedoc/dist/lib/converter/components';
import { RendererComponent } from 'typedoc/dist/lib/output/components';
import { PageEvent } from 'typedoc/dist/lib/output/events';

import { FrontMatter, Sidebar } from './types';

@Component({ name: 'front-matter' })
export class FrontMatterComponent extends RendererComponent {
  @BindOption('out')
  out!: string;
  @BindOption('sidebar')
  sidebar!: Sidebar;
  @BindOption('globalsTitle')
  globalsTitle!: string;
  @BindOption('readmeTitle')
  readmeTitle!: string;
  @BindOption('entryDocument')
  entryDocument!: string;

  globalsFile = 'modules.md';

  initialize() {
    super.initialize();
    this.listenTo(this.application.renderer, {
      [PageEvent.END]: this.onPageEnd,
    });
  }
  onPageEnd(page: PageEvent) {
    if (page.contents) {
      page.contents = prependYAML(page.contents, this.getYamlItems(page));
    }
  }

  getYamlItems(page: PageEvent): any {
    const pageId = this.getId(page);
    const pageTitle = this.getTitle(page);
    const sidebarLabel = this.getSidebarLabel(page);
    let items: FrontMatter = {
      id: pageId,
      title: pageTitle,
    };
    if (page.url === this.entryDocument) {
      items = { ...items, slug: '/' + this.out };
    }
    if (sidebarLabel && sidebarLabel !== pageTitle) {
      items = { ...items, sidebar_label: sidebarLabel };
    }
    if (page.url === page.project.url) {
      items = { ...items, hide_table_of_contents: true };
    }
    return {
      ...items,
      custom_edit_url: null,
      hide_title: true,
    };
  }

  getSidebarLabel(page: PageEvent) {
    if (!this.sidebar) {
      return null;
    }
    if (page.url === this.entryDocument) {
      return page.url === page.project.url
        ? this.sidebar.indexLabel
        : this.sidebar.readmeLabel;
    }

    if (page.url === this.globalsFile) {
      return this.sidebar.indexLabel;
    }
    return this.sidebar.fullNames ? page.model.getFullName() : page.model.name;
  }

  getId(page: PageEvent) {
    return path.basename(page.url, path.extname(page.url));
  }

  getTitle(page: PageEvent) {
    const readmeTitle = this.readmeTitle || page.project.name;
    if (page.url === this.entryDocument && page.url !== page.project.url) {
      return readmeTitle;
    }
    return getPageTitle(page);
  }
}
