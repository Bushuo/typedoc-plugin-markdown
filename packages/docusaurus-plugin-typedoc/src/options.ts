import {
  Application,
  MixedDeclarationOption,
  ParameterType,
  StringDeclarationOption,
  TSConfigReader,
  TypeDocReader,
} from 'typedoc';

import { PluginOptions } from './types';

/**
 * Default plugin options
 */
const DEFAULT_PLUGIN_OPTIONS: PluginOptions = {
  out: 'api',
  entryDocument: 'index.md',
  hideInPageTOC: true,
  hideBreadcrumbs: true,
  id: 'default',
  docsRoot: 'docs',
  sidebar: {
    fullNames: false,
    sidebarFile: 'typedoc-sidebar.js',
    indexLabel: 'Table of contents',
    readmeLabel: 'Readme',
  },
  readmeTitle: undefined,
};

/**
 * Merge default with user options
 * @param opts
 */
export const getOptions = (
  siteDir: string,
  opts: Partial<PluginOptions>,
): PluginOptions => ({
  ...DEFAULT_PLUGIN_OPTIONS,
  ...opts,
  ...(opts.sidebar && {
    sidebar: {
      ...DEFAULT_PLUGIN_OPTIONS.sidebar,
      ...opts.sidebar,
    },
  }),
  plugin: [
    ...['typedoc-plugin-markdown'],
    ...(opts.plugin
      ? opts.plugin.filter((name) => name !== 'typedoc-plugin-markdown')
      : []),
  ],
  siteDir,
});

/**
 * Add docusaurus options to converter
 * @param app
 */
export const addOptions = (app: Application) => {
  // configure deault typedoc options
  app.options.addReader(new TypeDocReader());
  app.options.addReader(new TSConfigReader());

  // expose plugin options to typedoc so we can access if required
  app.options.addDeclaration({
    name: 'id',
  } as StringDeclarationOption);

  app.options.addDeclaration({
    name: 'docsRoot',
  } as StringDeclarationOption);

  app.options.addDeclaration({
    name: 'siteDir',
  } as MixedDeclarationOption);

  app.options.addDeclaration({
    name: 'globalsTitle',
  } as StringDeclarationOption);

  app.options.addDeclaration({
    name: 'readmeTitle',
  } as StringDeclarationOption);

  app.options.addDeclaration({
    name: 'sidebar',
    type: ParameterType.Mixed,
  } as MixedDeclarationOption);
};
