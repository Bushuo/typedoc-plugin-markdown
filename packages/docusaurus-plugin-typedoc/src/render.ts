import { ProjectReflection, UrlMapping } from 'typedoc';
import { RendererEvent } from 'typedoc/dist/lib/output/events';
import * as ts from 'typescript';

export async function render(
  project: ProjectReflection,
  outputDirectory: string,
) {
  if (!this.prepareTheme() || !this.prepareOutputDirectory(outputDirectory)) {
    return;
  }

  const output = new RendererEvent(
    RendererEvent.BEGIN,
    outputDirectory,
    project,
  );

  output.settings = this.application.options.getRawValues();
  output.urls = this.theme!.getUrls(project);

  this.trigger(output);

  if (!output.isDefaultPrevented) {
    output.urls?.forEach((mapping: UrlMapping, i) => {
      this.renderDocument(output.createPageEvent(mapping));
      ts.sys.write(
        `\rGenerated ${i + 1} of ${output.urls?.length} TypeDoc docs`,
      );
    });
    ts.sys.write(`\n`);
    this.trigger(RendererEvent.END, output);
  }
}
