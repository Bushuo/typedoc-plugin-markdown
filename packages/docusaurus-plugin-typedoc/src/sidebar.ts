import * as fs from 'fs';
import * as path from 'path';

import { BindOption } from 'typedoc';
import { Component } from 'typedoc/dist/lib/converter/components';
import { RendererComponent } from 'typedoc/dist/lib/output/components';
import { RendererEvent } from 'typedoc/dist/lib/output/events';

import { SidebarItem, SidebarOptions } from './types';

@Component({ name: 'sidebar' })
export class SidebarComponent extends RendererComponent {
  @BindOption('sidebar')
  sidebar!: SidebarOptions;
  @BindOption('siteDir')
  siteDir!: string;
  @BindOption('out')
  out!: string;

  initialize() {
    this.listenTo(this.application.renderer, {
      [RendererEvent.BEGIN]: this.onRendererBegin,
    });
  }

  onRendererBegin(renderer: RendererEvent) {
    const navigation = this.application.renderer.theme?.getNavigation(
      renderer.project,
    );

    // map the navigation object to a Docuaurus sidebar format
    const sidebarItems = navigation?.children
      ? navigation.children.map((navigationItem) => {
          if (navigationItem.isLabel) {
            const sidebarCategoryItems = navigationItem.children
              ? navigationItem.children.map((navItem) => {
                  const url = this.getUrlKey(this.out, navItem.url);
                  if (navItem.children && navItem.children.length > 0) {
                    const sidebarCategoryChildren = navItem.children.map(
                      (childGroup) =>
                        this.getSidebarCategory(
                          childGroup.title,
                          childGroup.children
                            ? childGroup.children.map((childItem) =>
                                this.getUrlKey(this.out, childItem.url),
                              )
                            : [],
                        ),
                    );
                    return this.getSidebarCategory(navItem.title, [
                      url,
                      ...sidebarCategoryChildren,
                    ]);
                  }
                  return url;
                })
              : [];
            return this.getSidebarCategory(
              navigationItem.title,
              sidebarCategoryItems,
            );
          }
          return this.getUrlKey(this.out, navigationItem.url);
        })
      : [];

    writeSidebar(
      this.sidebar,
      `module.exports = ${JSON.stringify(sidebarItems, null, 2)};`,
    );

    this.application.logger.success(
      `TypeDoc sidebar written to ${this.sidebar.sidebarPath}`,
    );
  }

  /**
   * returns a sidebar category node
   */
  getSidebarCategory(title: string, items: SidebarItem[]) {
    return {
      type: 'category',
      label: title,
      items,
    };
  }

  /**
   * returns the url key for relevant doc
   */
  getUrlKey(out: string, url: string) {
    const urlKey = url.replace('.md', '');
    return out ? out + '/' + urlKey : urlKey;
  }
}

/**
 * Write content to sidebar file
 */
export const writeSidebar = (sidebar: SidebarOptions, content: string) => {
  if (!fs.existsSync(path.dirname(sidebar.sidebarPath))) {
    fs.mkdirSync(path.dirname(sidebar.sidebarPath));
  }
  fs.writeFileSync(sidebar.sidebarPath, content);
};
