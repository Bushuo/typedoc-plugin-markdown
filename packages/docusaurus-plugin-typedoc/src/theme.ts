import { ReflectionKind } from 'typedoc';
import MarkdownTheme from 'typedoc-plugin-markdown/dist/theme';
import { Renderer } from 'typedoc/dist/lib/output/renderer';

import { FrontMatterComponent } from './front-matter';

export default class DocusaurusTheme extends MarkdownTheme {
  constructor(renderer: Renderer, basePath: string) {
    super(renderer, basePath);
    renderer.addComponent('fm', new FrontMatterComponent(renderer));
  }

  allowedDirectoryListings() {
    return [...super.allowedDirectoryListings(), ...['_category_.yml']];
  }

  get mappings() {
    return super.mappings.map((mapping) => {
      if (mapping.kind.includes(ReflectionKind.Namespace)) {
        return {
          ...mapping,
          directory: 'namespaces',
        };
      }
      return mapping;
    });
  }
}
