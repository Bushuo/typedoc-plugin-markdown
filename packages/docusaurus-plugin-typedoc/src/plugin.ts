import * as path from 'path';

import { LoadContext } from '@docusaurus/types';
import { Application, TSConfigReader, TypeDocReader } from 'typedoc';

import { DocusaurusFrontMatterComponent } from './front-matter';
import { writeSidebar } from './sidebar';
import DocusaurusTheme from './theme/theme';
import { PluginOptions } from './types';

const DEFAULT_PLUGIN_OPTIONS: PluginOptions = {
  docsRoot: 'docs',
  out: 'api',
  sidebar: {
    fullNames: false,
    sidebarFile: 'typedoc-sidebar.js',
    globalsLabel: 'Index',
    readmeLabel: 'Readme',
  },
  globalsTitle: undefined,
  readmeTitle: undefined,
};

let app: Application;

export default async function pluginDocusaurus(
  context: LoadContext,
  opts: Partial<PluginOptions>,
) {
  const { siteDir } = context;

  /**
   * Configure options
   */
  const options = {
    ...DEFAULT_PLUGIN_OPTIONS,
    ...opts,
    ...(opts.sidebar && {
      sidebar: {
        ...DEFAULT_PLUGIN_OPTIONS.sidebar,
        ...opts.sidebar,
      },
    }),
  };

  /**
   * Initialize and build app
   * - currently we can't re-compile with devServer due to infinate render loop
   * - @TODO look if configureWebpack lifecycle can be updated to handle more gracefully
   */
  if (!app) {
    app = new Application();

    // TypeDoc options
    const typedocOptions = Object.keys(options).reduce((option, key) => {
      if (![...['id'], ...Object.keys(DEFAULT_PLUGIN_OPTIONS)].includes(key)) {
        option[key] = options[key];
      }
      return option;
    }, {});

    app.options.addReader(new TypeDocReader());
    app.options.addReader(new TSConfigReader());

    // bootstrap TypeDoc app
    app.bootstrap({
      // filtered TypeDoc options
      ...typedocOptions,
      // TypeDoc plugins
      plugin: [
        ...['typedoc-plugin-markdown'],
        ...(opts.plugin
          ? opts.plugin.filter((name) => name !== 'typedoc-plugin-markdown')
          : []),
      ],
      // add docusaurus theme
      theme: path.resolve(__dirname, 'theme'),
    });

    // add frontmatter component
    app.renderer.addComponent(
      'docusaurus-frontmatter',
      new DocusaurusFrontMatterComponent(app.renderer, options),
    );

    // return the generated reflections
    const project = app.convert();

    // if project is undefined typedoc has a problem - error logging will be supplied by typedoc.
    if (!project) {
      return;
    }

    // construct outputDirectory path
    const outputDirectory = path.resolve(
      siteDir,
      options.docsRoot,
      options.out,
    );

    // generate the static docs
    await app.generateDocs(project, outputDirectory);

    // write the sidebar (if applicable)
    if (options.sidebar) {
      const theme = app.renderer.getComponent('theme') as DocusaurusTheme;
      writeSidebar(
        options.disableOutputCheck || theme.isOutputDirectory(outputDirectory),
        siteDir,
        options.out,
        options.sidebar,
        theme.getNavigation(project),
      );
    }
  }

  // we need to generate the sidebar before any available lifecycle apis
  return {
    name: 'docusaurus-plugin-typedoc',
  };
}
