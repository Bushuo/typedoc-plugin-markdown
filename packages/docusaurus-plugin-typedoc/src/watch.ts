import * as fs from 'fs';
import * as path from 'path';

import { Application } from 'typedoc';

import { getOutputDirectory } from './options';
import { getSidebarPath } from './sidebar';
import { PluginOptions } from './types';

/**
 * Calls TypeDoc's `convertAndWatch` and force trigger sidebars refresh.
 */
export const convertAndWatch = (app: Application, options: PluginOptions) => {
  const sidebarsJsPath = path.resolve(options.siteDir, 'sidebars.js');
  const sidebarPath = getSidebarPath(options);
  app.convertAndWatch(async (project) => {
    if (sidebarPath) {
      // remove typedoc sidebar from require cache
      delete require.cache[sidebarPath];
      // force trigger a sidebars.js refresh
      const sidebarJsContent = fs.readFileSync(sidebarsJsPath);
      fs.writeFileSync(sidebarsJsPath, sidebarJsContent);
    }
    await app.generateDocs(project, getOutputDirectory(options));
  });
};
