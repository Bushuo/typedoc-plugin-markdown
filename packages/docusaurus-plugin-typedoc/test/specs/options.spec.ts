import { getOptions } from '../../src/options';
import { PluginOptions } from '../../src/types';

describe(`Options:`, () => {
  test(`should return default options`, () => {
    expect(
      getOptions('site-dir', { entryPoints: ['entrypoint.ts'] }),
    ).toMatchSnapshot();
  });
  test(`should merge plugins`, () => {
    expect(
      getOptions('site-dir', {
        plugin: ['plugin-xyz'],
      }).plugin,
    ).toMatchSnapshot();
  });
  test(`should not duplicate markdown plugin`, () => {
    expect(
      getOptions('site-dir', {
        plugin: ['typedoc-plugin-markdown', 'plugin-xyz'],
      }).plugin,
    ).toMatchSnapshot();
  });
  test(`should merge sidebar options`, () => {
    expect(
      getOptions('site-dir', {
        sidebar: { sidebarFile: 'custom-path.js', fullNames: true },
      } as PluginOptions).sidebar,
    ).toMatchSnapshot();
  });
});
