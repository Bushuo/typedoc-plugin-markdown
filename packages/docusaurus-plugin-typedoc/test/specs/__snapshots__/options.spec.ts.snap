// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Options: should merge custom sidebar options 1`] = `
Object {
  "fullNames": true,
  "indexLabel": "Custom index",
  "readmeLabel": "Custom readme",
  "sidebarFile": "custom-sidebar.js",
  "sidebarPath": "/site-dir/custom-sidebar.js",
}
`;

exports[`Options: should merge plugins 1`] = `
Array [
  "plugin-xyz",
]
`;

exports[`Options: should return default options 1`] = `
Object {
  "docsRoot": "docs",
  "entryDocument": "index.md",
  "entryPoints": Array [
    "entrypoint.ts",
  ],
  "hideBreadcrumbs": true,
  "hideInPageTOC": true,
  "id": "default",
  "out": "api",
  "outputDirectory": "/site-dir/docs/api",
  "plugin": Array [
    "none",
  ],
  "sidebar": Object {
    "fullNames": false,
    "indexLabel": "Table of contents",
    "readmeLabel": "Readme",
    "sidebarFile": "typedoc-sidebar.js",
    "sidebarPath": "/site-dir/typedoc-sidebar.js",
  },
  "siteDir": "/site-dir",
  "watch": false,
}
`;
